storagemarket/contract_deal_monitor.go

```diff
@@ -0,0 +1,269 @@
+
+					fromEthAddr, err := ethtypes.ParseEthAddress(_from)
+					if err != nil {
+						return fmt.Errorf("parsing `from` eth address failed: %w", err)
+					}
+
+					toEthAddr, err := ethtypes.ParseEthAddress(_to)
+					if err != nil {
+						return fmt.Errorf("parsing `to` eth address failed: %w", err)
+					}
+
+					params, err := ethtypes.DecodeHexString(_params)
+					if err != nil {
+						return fmt.Errorf("decoding params failed: %w", err)
+					}
+
+					res, err := c.api.EthCall(ctx, ethtypes.EthCall{
+						From: &fromEthAddr,
+						To:   &toEthAddr,
+						Data: params,
+					}, "latest")
+					if err != nil {
+						return fmt.Errorf("eth call erred: %w", err)
+					}
+
+					begin, length, err := lengthPrefixPointsTo(res)
+					if err != nil {
+						return fmt.Errorf("length prefix points erred: %w", err)
+					}
+
+					var dpc types.ContractDealProposal
+					err = dpc.UnmarshalCBOR(bytes.NewReader(res[begin : begin+length]))
+					if err != nil {
+						return fmt.Errorf("cbor unmarshal failed: %w", err)
+					}
+
+					var pv1 types.ParamsVersion1
+					err = pv1.UnmarshalCBOR(bytes.NewReader(dpc.Params))
+					if err != nil {
+						return fmt.Errorf("params cbor unmarshal failed: %w", err)
+					}
+
+					rootCidStr, err := dpc.Label.ToString()
+					if err != nil {
+						return fmt.Errorf("getting cid from label failed: %w", err)
+					}
+
+					rootCid, err := cid.Parse(rootCidStr)
+					if err != nil {
+						return fmt.Errorf("parsing cid failed: %w", err)
+					}
+
+					prop := market.DealProposal{
+						PieceCID:     dpc.PieceCID,
+						PieceSize:    dpc.PieceSize,
+						VerifiedDeal: dpc.VerifiedDeal,
+						Client:       dpc.Client,
+						Provider:     c.maddr,
+
+						Label: dpc.Label,
+
+						StartEpoch:           dpc.StartEpoch,
+						EndEpoch:             dpc.EndEpoch,
+						StoragePricePerEpoch: dpc.StoragePricePerEpoch,
+
+						ProviderCollateral: dpc.ProviderCollateral,
+						ClientCollateral:   dpc.ClientCollateral,
+					}
+
+					proposal := types.DealParams{
+						DealUUID:  uuid.New(),
+						IsOffline: false,
+						ClientDealProposal: market.ClientDealProposal{
+							Proposal: prop,
+							// signature is garbage, but it still needs to serialize, so shouldnt be empty!!
+							ClientSignature: crypto.Signature{
+								Type: crypto.SigTypeBLS,
+								Data: []byte{0xde, 0xad},
+							},
+						},
+						DealDataRoot: rootCid,
+						Transfer: types.Transfer{
+							Type:   "http",
+							Params: []byte(fmt.Sprintf(`{"URL":"%s"}`, pv1.LocationRef)),
+							Size:   pv1.CarSize,
+						},
+						//TODO: maybe add to pv1?? RemoveUnsealedCopy: paramsAndVersion.RemoveUnsealedCopy,
+						RemoveUnsealedCopy: false,
+						SkipIPNIAnnounce:   pv1.SkipIpniAnnounce,
+					}
+
+					log.Infow("received contract deal proposal", "id", proposal.DealUUID, "client-peer", dpc.Client)
+
+					reason, err := c.prov.ExecuteLibp2pDeal(context.Background(), &proposal, "")
```

Comment:
nit: We can probably leave the name as ExecuteDeal as it's called from the contract deal monitor path as well as from libp2p